doc/memory.rst

```diff
@@ -324,26 +324,65 @@ Gotchas
     >>> print(sin(0))
     0.0
 
-* **caching methods**: you cannot decorate a method at class definition,
-  because when the class is instantiated, the first argument (self) is
-  *bound*, and no longer accessible to the `Memory` object. The following
-  code won't work::
+* **caching methods: memory is designed for pure functions and it is
+  not recommended to use it for methods**. If you want to use cache
+  inside a class the recommended pattern is to cache a pure function
+  and use the cached function inside your class, i.e. something like
+  this::
+
+    @mem.cache
+    def compute_func(arg1, arg2, arg3):
+        # long computation
+        return result
+
 
     class Foo(object):
+        def __init__(self, args):
+            self.data = None
 
-        @mem.cache  # WRONG
-        def method(self, args):
-	    pass
+        def compute(self):
```

Comment:
I'll merge this PR and get rid of the mem.cache decorators in a separate PR.

Out of interest why do you do it this way to make it clearer that the function is cached where it is called, to keep the uncached function easily accessible, for some other reasons?